src/pages/files.tsx

```diff
@@ -298,6 +380,77 @@ function FilesApp(props: IFilesAppProps) {
     return '';
   }
 
+  /**
+   * Rename handler: prompts for a new name and performs a contents rename.
+   * Shows an error dialog if the rename fails, for various reasons:
+   * - the new name is invalid (empty, contains slashes)
+   * - a file with the new name already exists
+   * - the contents manager fails to perform the rename for whatever reason
+   */
+  const renameFile = React.useCallback(
+    async (model: Contents.IModel) => {
+      const oldName = model.name;
+      let attempt = oldName; // this is what we prefill if re-opening the dialog.
+
+      // We'll loop until the user cancels, or we complete a valid rename
+      // eslint-disable-next-line no-constant-condition
+      while (true) {
+        const result = await openRenameDialog(attempt);
+
+        if (!result.button.accept) {
+          return;
+        }
+
+        const newName = (result.value?.newName ?? '').trim();
+        attempt = newName;
+
+        // If there is no change, this is a no-op, so we just quietly return.
+        if (!newName || newName === oldName) {
+          return;
+        }
+
+        if (/[\\/]/.test(newName)) {
+          await showErrorMessage(
+            'Invalid name',
+            'File name cannot contain “/” or “\\”. Please choose a different name.'
+          );
+          continue;
+        }
+
+        const oldExt = oldName.includes('.') ? oldName.split('.').pop()! : '';
+        let finalName = newName;
+        if (oldExt && !newName.includes('.')) {
+          finalName = `${newName}.${oldExt}`;
+        }
+
+        const dirname = model.path.split('/').slice(0, -1).join('/');
```

Comment:
You could use `PathExt` from lumino since this is already installed and may handle some edge cases better.

Reply:
Sounds good!